src/Compilers/CSharp/Portable/Parser/LanguageParser_Patterns.cs

```diff
@@ -536,6 +538,13 @@ SyntaxKind.GreaterThanToken or
             if (@this.CurrentToken.Kind is SyntaxKind.CloseParenToken or SyntaxKind.CloseBraceToken or SyntaxKind.CloseBracketToken or SyntaxKind.SemicolonToken)
                 return PostSkipAction.Abort;
 
+            // `:` is usually treated as incorrect separation token. This helps for error recovery in basic typing scenarios like `{ Prop:$$ Prop1: { ... } }`.
+            // However, such behavior isn't much desirable when parsing pattern of a case label in a switch statement. For instance, consider the following example: `case { Prop: { }: case ...`.
+            // Normally we would skip second `:` and `case` keyword after it as bad tokens and continue parsing pattern, which produces a lot of noise errors.
+            // In order to avoid that and produce single error of missing `}` we exit on unexpected `:` in such cases.
```

Comment:
Awesome description of the problem.

Reply:
nit: wrapping at 120 columns tends to help for prs.